src/lib/search-params.ts

```diff
@@ -0,0 +1,150 @@
+  },
+};
+
+interface ClientSearchOptions {
+  query?: SearchOptions['query'];
+  offset?: SearchOptions['offset'];
+  filters: {
+    licenses?: string[];
+    publishers?: string[];
+  };
+  sortBy?: SortBy;
+}
+
+export function getUrlWithSearchParams({
+  query,
+  offset,
+  filters: {licenses, publishers},
+  sortBy,
+}: ClientSearchOptions): string {
+  const searchParams: {[key: string]: string} = {};
+
+  if (query) {
+    searchParams.query = query;
+  }
+
+  if (licenses?.length) {
+    searchParams.licenses = licenses.join(',');
+  }
+
+  if (publishers?.length) {
+    searchParams.publishers = publishers.join(',');
+  }
+
+  if (offset) {
+    searchParams.offset = `${offset}`;
+  }
+
+  if (sortBy && sortBy !== defaultSortBy) {
+    searchParams.sortBy = sortBy;
+  }
+
+  const encodedSearchParams = new URLSearchParams(searchParams).toString();
+
+  if (encodedSearchParams) {
+    return '/?' + encodedSearchParams;
+  } else {
+    return '/' + encodedSearchParams;
+  }
+}
+
+export interface SearchParams {
+  publishers?: string;
+  licenses?: string;
+  query?: string;
+  offset?: string;
+  sortBy?: SortBy;
+}
+
+// Based on https://github.com/colinhacks/zod/issues/316#issuecomment-1024793482
+export function fallback<T>(value: T) {
+  return any().transform(() => value);
+}
+
+// Always return a valid SearchOptions object, even if the search params aren't correct,
+// so the application doesn't fail on invalid search params.
+const searchOptionsWithFallbackSchema = searchOptionsSchema.extend({
+  offset: searchOptionsSchema.shape.offset.or(fallback(0)),
+  sortBy: searchOptionsSchema.shape.sortBy.or(
+    fallback(SortBySearchOption.Relevance)
+  ),
+  sortOrder: searchOptionsSchema.shape.sortOrder.or(
+    fallback(SortOrder.Descending)
+  ),
+  query: z.string().optional(), // Don't default to "*"
+});
+
+// Always return the sort values so the client knows how to sort.
+interface SearchOptionsWithRequiredSort extends SearchOptions {
+  sortBy: NonNullable<SearchOptions['sortBy']>;
+  sortOrder: NonNullable<SearchOptions['sortOrder']>;
+}
+
+// This function translates the search params to valid search options.
+// Next.js already separate the search query string into separates properties with string values.
+export function fromSearchParamsToSearchOptions({
+  publishers,
+  licenses,
+  query,
+  offset = '0',
+  sortBy = defaultSortBy,
+}: SearchParams): SearchOptionsWithRequiredSort {
+  const {sortBy: sortBySearchOption, sortOrder} = sortMapping[sortBy] || {};
+
+  // Transform the string values from the search params to SearchOptions
+  const options = {
+    offset: +offset,
+    filters: {
+      publishers: publishers?.split(',').filter(id => !!id),
+      licenses: licenses?.split(',').filter(id => !!id),
+    },
+    sortBy: sortBySearchOption,
+    sortOrder: sortOrder,
+    query: query,
+  };
+
+  const validOptions: SearchOptionsWithRequiredSort =
+    searchOptionsWithFallbackSchema.parse(options);
```

Comment:
Nice! A thought (not required for now!): would it be possible to make the transformation in `const options` also a part of the Zod schema (e.g. converting the offset string to a number, splitting the publishers string into an array)? Then we'd have all validation and transformation in one place?